<a href="https://colab.research.google.com/github/sergiocostaifes/PPCOMP_DM/blob/main/notebooks/07_baseline_rf.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# 07_baseline_rf.ipynb — Baseline Supervisionado (Random Forest)

## Objetivo

Treinar e avaliar um baseline supervisionado com **Random Forest** sobre a base rotulada em janelas de 5 minutos (Notebook 06), considerando a definição de episódios críticos baseada em **taxa de falha** (Notebooks 04 e 05).

Serão avaliados dois cenários:

1. **Classificação binária**  
   `is_event = 1` para `{BEFORE, DURING, AFTER}` vs `NORMAL`

2. **Classificação multiclasse**  
   `{NORMAL, BEFORE, DURING, AFTER}`

A avaliação adota **split temporal** para evitar vazamento de informação futura.

## Contexto metodológico (coerência com a taxa)

- Os episódios críticos foram detectados no Notebook 04 usando `fail_rate = n_failed / n_events` e limiar `μ + 2σ`.
- As features foram construídas no Notebook 05 **sobre `fail_rate`**, incluindo lags, rolling (1h), variação percentual e z-score global.
- A rotulagem temporal BEFORE/DURING/AFTER foi gerada no Notebook 06 com precedência `DURING > BEFORE/AFTER > NORMAL`.

Este notebook não redefine episódios nem rótulos; apenas treina e avalia modelos a partir dos artefatos persistidos.

## Entradas (artefatos do pipeline)

- `window_5min_labeled.parquet` (Notebook 06)

## Saídas (artefatos deste notebook)

- `rf_binary.joblib`
- `rf_multiclass.joblib`
- `07_baseline_rf_summary.json`

## Split temporal

Este notebook suporta duas estratégias:

- **Corte fixo (default)**: treino = 80% inicial; teste = 20% final
- **TimeSeriesSplit (opcional)**: validação em múltiplos folds temporais

## Métricas reportadas

- **Binário**: accuracy, precision, recall, f1, ROC-AUC (quando aplicável), matriz de confusão  
- **Multiclasse**: accuracy, macro-f1, weighted-f1, classification report, matriz de confusão

## Seleção de features e prevenção de vazamento

- Remove colunas de rótulo (`state`) e identificadores (`bucket_id`)
- Remove `is_critical` quando presente, pois deriva diretamente da detecção do NB04 (pode induzir vazamento conceitual)
- Mantém apenas colunas numéricas como features
- Observação: `bucket_start_us` é numérica, mas representa o tempo absoluto; será mantida por padrão como baseline, podendo ser removida em iterações futuras para testar invariância temporal

## Observação

O dataset é desbalanceado, especialmente em DURING. O Random Forest é utilizado como baseline interpretável e robusto, com `class_weight` ativado para mitigar desequilíbrio.

In [1]:
# ============================================================
# 07_baseline_rf.ipynb
# Baseline supervisionado com Random Forest (binário vs multiclasse)
# Split temporal + métricas + matriz de confusão + salvamento do modelo
# Pipeline PPCOMP_DM (Google Cluster Trace) - coerente com taxa (fail_rate)
# ============================================================

# -----------------------------
# 0) BOOTSTRAP (Colab + Repo)
# -----------------------------
from pathlib import Path
import os
import sys
import subprocess
import importlib
import random
import numpy as np
import pandas as pd
import json

SEED = 42
random.seed(SEED)
np.random.seed(SEED)

if not Path("/content/drive/MyDrive").exists():
    from google.colab import drive
    drive.mount("/content/drive")
else:
    print("[Bootstrap] Google Drive já montado.")

REPO_DIR = Path("/content/drive/MyDrive/Mestrado/PPCOMP_DM")
GITHUB_REPO = "https://github.com/sergiocostaifes/PPCOMP_DM.git"

if not REPO_DIR.exists():
    REPO_DIR.parent.mkdir(parents=True, exist_ok=True)
    print(f"[Bootstrap] Clonando repositório em: {REPO_DIR}")
    subprocess.run(["git", "clone", GITHUB_REPO, str(REPO_DIR)], check=True)
else:
    try:
        print("[Bootstrap] Atualizando repositório (git pull).")
        subprocess.run(["git", "-C", str(REPO_DIR), "pull"], check=True)
    except Exception as e:
        print("[Bootstrap] Aviso:", e)

os.chdir(str(REPO_DIR))
print("[Bootstrap] CWD =", os.getcwd())

repo_str = str(REPO_DIR)
if repo_str not in sys.path:
    sys.path.insert(0, repo_str)
importlib.invalidate_caches()

from src.paths import FEATURES_PATH, REPORTS_PATH, MODELS_PATH, ensure_dirs
ensure_dirs()

print("FEATURES_PATH =", FEATURES_PATH)
print("REPORTS_PATH  =", REPORTS_PATH)
print("MODELS_PATH   =", MODELS_PATH)

def log(msg: str) -> None:
    print(f"[07_baseline_rf] {msg}")

# -----------------------------
# 1) Imports ML
# -----------------------------
from sklearn.model_selection import TimeSeriesSplit
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import (
    accuracy_score,
    precision_recall_fscore_support,
    classification_report,
    confusion_matrix,
    roc_auc_score
)
import joblib

# -----------------------------
# 2) Carregar dataset rotulado (NB06)
# -----------------------------
DATA_FILE = FEATURES_PATH / "window_5min_labeled.parquet"
assert DATA_FILE.exists(), f"Arquivo não encontrado: {DATA_FILE}"

df = pd.read_parquet(DATA_FILE).sort_values("bucket_id").reset_index(drop=True)

assert "state" in df.columns, "Coluna 'state' ausente."
assert "bucket_id" in df.columns, "Coluna 'bucket_id' ausente."
assert "fail_rate" in df.columns, "Coluna 'fail_rate' ausente (pipeline por taxa não está refletido)."

log(f"Dataset rotulado: shape={df.shape}")
dist_total = df["state"].value_counts().to_dict()
log(f"Distribuição total: {dist_total}")

# -----------------------------
# 3) Preparar features (X) e alvos (y)
# -----------------------------
REMOVE_TIME_ABSOLUTE = False  # se True, remove bucket_start_us das features

drop_cols = {"state", "bucket_id"}

# is_critical é derivado do NB04; remover para evitar vazamento conceitual
if "is_critical" in df.columns:
    drop_cols.add("is_critical")

if REMOVE_TIME_ABSOLUTE and "bucket_start_us" in df.columns:
    drop_cols.add("bucket_start_us")

numeric_cols = df.select_dtypes(include=[np.number]).columns.tolist()
X_cols = [c for c in numeric_cols if c not in drop_cols]

if len(X_cols) == 0:
    raise ValueError("Nenhuma feature numérica encontrada após filtragem.")

X = df[X_cols].copy()
y_multi = df["state"].astype(str).copy()

log(f"Features usadas: {len(X_cols)}")
log(f"Lista features: {X_cols}")

# -----------------------------
# 4) Split temporal (corte fixo default)
# -----------------------------
SPLIT_MODE = "fixed"   # "fixed" ou "tscv"
TEST_RATIO = 0.20

n = len(df)
test_size = int(np.ceil(n * TEST_RATIO))
train_end = n - test_size

X_train, X_test = X.iloc[:train_end], X.iloc[train_end:]
y_train_multi, y_test_multi = y_multi.iloc[:train_end], y_multi.iloc[train_end:]

log(f"Split temporal: train={len(X_train)} test={len(X_test)} (modo={SPLIT_MODE})")

# distribuição por split (útil para validar desbalanceamento)
dist_train = y_train_multi.value_counts().to_dict()
dist_test = y_test_multi.value_counts().to_dict()
log(f"Distribuição treino: {dist_train}")
log(f"Distribuição teste : {dist_test}")

# -----------------------------
# 5) Cenário binário: evento vs normal
# -----------------------------
y_train_bin = (y_train_multi != "NORMAL").astype(int)
y_test_bin = (y_test_multi != "NORMAL").astype(int)

rf_bin = RandomForestClassifier(
    n_estimators=300,
    random_state=SEED,
    n_jobs=-1,
    class_weight="balanced_subsample"
)
rf_bin.fit(X_train, y_train_bin)

pred_bin = rf_bin.predict(X_test)
proba_bin = rf_bin.predict_proba(X_test)[:, 1] if hasattr(rf_bin, "predict_proba") else None

acc_bin = accuracy_score(y_test_bin, pred_bin)
prec_bin, rec_bin, f1_bin, _ = precision_recall_fscore_support(
    y_test_bin, pred_bin, average="binary", zero_division=0
)

auc_bin = None
if proba_bin is not None and len(np.unique(y_test_bin)) == 2:
    try:
        auc_bin = float(roc_auc_score(y_test_bin, proba_bin))
    except Exception:
        auc_bin = None

cm_bin = confusion_matrix(y_test_bin, pred_bin).tolist()
log(f"[BIN] acc={acc_bin:.4f} prec={prec_bin:.4f} rec={rec_bin:.4f} f1={f1_bin:.4f} auc={auc_bin}")

# -----------------------------
# 6) Cenário multiclasse: NORMAL/BEFORE/DURING/AFTER
# -----------------------------
labels = ["NORMAL", "BEFORE", "DURING", "AFTER"]  # ordem fixa para matriz comparável

rf_multi = RandomForestClassifier(
    n_estimators=400,
    random_state=SEED,
    n_jobs=-1,
    class_weight="balanced_subsample"
)
rf_multi.fit(X_train, y_train_multi)

pred_multi = rf_multi.predict(X_test)

acc_multi = accuracy_score(y_test_multi, pred_multi)

macro_prec, macro_rec, macro_f1, _ = precision_recall_fscore_support(
    y_test_multi, pred_multi, average="macro", zero_division=0
)
w_prec, w_rec, w_f1, _ = precision_recall_fscore_support(
    y_test_multi, pred_multi, average="weighted", zero_division=0
)

cm_multi = confusion_matrix(y_test_multi, pred_multi, labels=labels).tolist()
report_multi = classification_report(y_test_multi, pred_multi, labels=labels, zero_division=0)

log(f"[MULTI] acc={acc_multi:.4f} macro_f1={macro_f1:.4f} weighted_f1={w_f1:.4f}")

# -----------------------------
# 7) (Opcional) TimeSeriesSplit rápido (apenas multiclasse)
# -----------------------------
tscv_results = []
if SPLIT_MODE == "tscv":
    tscv = TimeSeriesSplit(n_splits=5)
    for fold, (tr, te) in enumerate(tscv.split(X), start=1):
        X_tr, X_te = X.iloc[tr], X.iloc[te]
        y_tr, y_te = y_multi.iloc[tr], y_multi.iloc[te]

        clf = RandomForestClassifier(
            n_estimators=300,
            random_state=SEED,
            n_jobs=-1,
            class_weight="balanced_subsample"
        )
        clf.fit(X_tr, y_tr)
        y_hat = clf.predict(X_te)

        f1_macro = precision_recall_fscore_support(
            y_te, y_hat, average="macro", zero_division=0
        )[2]
        tscv_results.append({"fold": fold, "f1_macro": float(f1_macro)})

# -----------------------------
# 8) Salvar modelos
# -----------------------------
bin_model_path = MODELS_PATH / "rf_binary.joblib"
multi_model_path = MODELS_PATH / "rf_multiclass.joblib"

joblib.dump(rf_bin, bin_model_path)
joblib.dump(rf_multi, multi_model_path)

# -----------------------------
# 9) Summary (JSON)
# -----------------------------
summary = {
    "seed": SEED,
    "split_mode": SPLIT_MODE,
    "test_ratio": TEST_RATIO,
    "rows_total": int(n),
    "rows_train": int(len(X_train)),
    "rows_test": int(len(X_test)),
    "n_features": int(len(X_cols)),
    "features": X_cols,
    "class_distribution_total": {k: int(v) for k, v in dist_total.items()},
    "class_distribution_train": {k: int(v) for k, v in dist_train.items()},
    "class_distribution_test": {k: int(v) for k, v in dist_test.items()},
    "binary": {
        "acc": float(acc_bin),
        "precision": float(prec_bin),
        "recall": float(rec_bin),
        "f1": float(f1_bin),
        "roc_auc": auc_bin,
        "confusion_matrix": cm_bin
    },
    "multiclass": {
        "acc": float(acc_multi),
        "macro_f1": float(macro_f1),
        "weighted_f1": float(w_f1),
        "confusion_matrix": cm_multi,
        "labels_order": labels,
        "classification_report_text": report_multi
    },
    "tscv_multiclass": tscv_results,
    "models": {
        "binary_path": str(bin_model_path),
        "multiclass_path": str(multi_model_path)
    }
}

summary_file = REPORTS_PATH / "07_baseline_rf_summary.json"
summary_file.write_text(json.dumps(summary, indent=2, ensure_ascii=False))

log("Notebook 07 finalizado com sucesso.")

print("\n=== MULTICLASS REPORT ===\n")
print(report_multi)
print("\n=== CONFUSION MATRIX (BIN) ===\n", cm_bin)
print("\n=== CONFUSION MATRIX (MULTI) labels=", labels, "===\n", cm_multi)
print("\nModels saved:\n", bin_model_path, "\n", multi_model_path)
print("\nSummary saved:\n", summary_file)

Mounted at /content/drive
[Bootstrap] Atualizando repositório (git pull).
[Bootstrap] CWD = /content/drive/MyDrive/Mestrado/PPCOMP_DM
FEATURES_PATH = /content/drive/MyDrive/Mestrado/02-datasets/03-features
REPORTS_PATH  = /content/drive/MyDrive/Mestrado/04-reports
MODELS_PATH   = /content/drive/MyDrive/Mestrado/03-models
[07_baseline_rf] Dataset rotulado: shape=(8914, 28)
[07_baseline_rf] Distribuição total: {'NORMAL': 5533, 'BEFORE': 1996, 'AFTER': 991, 'DURING': 394}
[07_baseline_rf] Features usadas: 25
[07_baseline_rf] Lista features: ['bucket_start_us', 'n_events', 'n_failed', 'n_machines', 'n_collections', 'mean_priority', 'mean_req_cpus', 'mean_req_mem', 'req_cpus_presence_rate', 'req_mem_presence_rate', 'event_FAIL_count', 'event_SCHEDULE_count', 'event_FINISH_count', 'event_ENABLE_count', 'event_LOST_count', 'event_EVICT_count', 'event_KILL_count', 'fail_rate', 'lag_1', 'lag_2', 'lag_3', 'rolling_mean_1h', 'rolling_std_1h', 'pct_change', 'zscore_global']
[07_baseline_rf] Split 

## Achados do Notebook 07 — Baseline Supervisionado (Random Forest)

### Estrutura Experimental

- Total de janelas: 8914
- Split temporal fixo (80/20) preservando ordem cronológica
- 25 features numéricas utilizadas
- Dois cenários avaliados:
  - Classificação binária (evento vs NORMAL)
  - Classificação multiclasse (NORMAL / BEFORE / DURING / AFTER)
- Não houve vazamento temporal nem uso de colunas derivadas do rótulo.

---

### 1. Resultados — Classificação Binária

Métricas:

- Accuracy: 0.815
- Precision: 0.850
- Recall: 0.910
- F1-score: 0.879
- ROC-AUC: 0.889

Interpretação:

O modelo apresenta alta capacidade de detecção de eventos (recall ≈ 91%), com separabilidade estatística robusta (AUC ≈ 0.89).

Comparado à versão anterior baseada em volume absoluto, houve melhora substancial no recall e no F1-score, indicando que a definição de criticidade baseada em taxa produz sinal mais consistente para aprendizado supervisionado.

---

### 2. Resultados — Classificação Multiclasse

Métricas globais:

- Accuracy: 0.656
- Macro-F1: 0.599
- Weighted-F1: 0.608

Desempenho por classe:

- NORMAL: F1 ≈ 0.63
- BEFORE: F1 ≈ 0.69
- DURING: F1 = 1.00
- AFTER: F1 ≈ 0.07

---

### 3. Principais Achados

#### 3.1 Separação perfeita da classe DURING

A classe DURING foi identificada com precisão e recall iguais a 1.00.

Isso confirma que:

- A definição formal de episódios baseada em fail_rate é estatisticamente consistente.
- A engenharia de atributos captura adequadamente o regime crítico.
- Eventos críticos são fortemente separáveis no espaço de features.

Este é um resultado estrutural relevante para a dissertação.

---

#### 3.2 Forte melhoria na detecção de BEFORE

Diferentemente da versão anterior (volume absoluto), a classe BEFORE passou a apresentar:

- Recall elevado (~84%)
- F1 consistente (~0.69)

Isso indica que a abordagem por taxa torna a fase pré-evento estatisticamente modelável, fortalecendo o potencial preditivo do sistema.

---

#### 3.3 Dificuldade na classe AFTER

A classe AFTER apresenta F1 baixo (~0.07), com alta confusão com BEFORE.

Isso sugere que:

- A fase de estabilização pode compartilhar características estatísticas com a fase pré-evento.
- Pode ser necessário ajustar K_AFTER ou incorporar features temporais mais longas.
- Modelos sequenciais (LSTM, Transformers temporais) podem capturar melhor essa transição.

---

### 4. Implicações Científicas

1. A definição de episódios baseada em μ + 2σ sobre fail_rate mostrou-se operacionalmente robusta.
2. Eventos críticos são claramente separáveis.
3. A fase pré-evento tornou-se detectável sob a abordagem por taxa.
4. A fase pós-evento requer modelagem temporal mais sofisticada.
5. O baseline Random Forest estabelece referência comparativa para modelos avançados.

---

### 5. Conclusão

O Notebook 07 confirma que o pipeline completo (NB04–NB06) produz estrutura supervisionada consistente e cientificamente válida.

A partir deste ponto, a pesquisa pode evoluir para:

- Modelos temporais mais avançados
- Ajuste fino de hiperparâmetros
- Técnicas de balanceamento
- Avaliação preditiva antecipada de episódios